CovertInfiltration/Src/CovertInfiltration/Classes/UISSManager_CovertAction.uc

```diff
@@ -56,15 +58,20 @@ simulated protected function BuildConfiguration()
 	Configuration = new class'SSAAT_SquadSelectConfiguration';
 	History = `XCOMHISTORY;
 
-	Slots.Length = GetAction().StaffSlots.Length;
+	CovertAction = CovertOpsScreen.GetAction();
+	
+	Slots.Length = CovertAction.StaffSlots.Length;
+
 	for (i = 0; i < Slots.Length; ++i)
 	{
-		StaffSlotState = XComGameState_StaffSlot(History.GetGameStateForObjectID(GetAction().StaffSlots[i].StaffSlotRef.ObjectID));
-		RewardState = XComGameState_Reward(History.GetGameStateForObjectID(GetAction().StaffSlots[i].RewardRef.ObjectID));
+		StaffSlotState = XComGameState_StaffSlot(History.GetGameStateForObjectID(CovertAction.StaffSlots[i].StaffSlotRef.ObjectID));
+		RewardState = XComGameState_Reward(History.GetGameStateForObjectID(CovertAction.StaffSlots[i].RewardRef.ObjectID));
 
-		// Add notes. TODO: Add required class and rank notes
 		if (RewardState != none) Slots[i].Notes.AddItem(ConvertRewardToNote(RewardState));
-		if (GetAction().StaffSlots[i].bOptional) Slots[i].Notes.AddItem(CreateOptionalNote());
+		if (CovertAction.StaffSlots[i].bOptional) Slots[i].Notes.AddItem(CreateOptionalNote());
+		if (StaffSlotState.RequiredClass != '') Slots[i].Notes.AddItem(CreateClassNote(StaffSlotState.RequiredClass));
+		if (StaffSlotState.RequiredMinRank > 0) Slots[i].Notes.AddItem(CreateRankNote(StaffSlotState.RequiredMinRank, StaffSlotState.RequiredClass));
```

Comment:
What happens if there is no required class?

Reply:
Check `GetRankName` in `X2ExperienceConfig`. No class name (or a class name without explicit ranks) makes it fall back to the default ranks. The original covert ops code actually forgot to pass it, see `X2StrategyElement_XpackStaffSlots:GetCovertActionSoldierNameDisplayString`.